src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/Option.cs

```diff
@@ -0,0 +1,302 @@
+        }
+
+        /// <summary>
+        /// Perform one of two actions based on this option
+        /// </summary>
+        /// <param name="none">Action to perform for None{T}."/></param>
+        /// <param name="some">Action to perform for Some{T}.</param>
+        public void Match(Action none, Action<T> some)
+        {
+            if (none == null)
+            {
+                throw new ArgumentNullException(nameof(none));
+            }
+
+            if (some == null)
+            {
+                throw new ArgumentNullException(nameof(some));
+            }
+
+            some(Value);
+        }
+
+        /// <summary>
+        /// Act when this option is None{T}
+        /// </summary>
+        /// <param name="action">Action to invoke.</param>
+        /// <returns>This option, for chaining.</returns>
+        public IOption<T> WhenNone(Action action)
+        {
+            if (action == null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Act when this option is Some{T}
+        /// </summary>
+        /// <param name="action">Action to invoke.</param>
+        /// <returns>This option, for chaining.</returns>
+        public IOption<T> WhenSome(Action<T> action)
+        {
+            if (action == null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            action(Value);
+            return this;
+        }
+
+        /// <summary>
+        /// Return the value we wrap
+        /// </summary>
+        /// <param name="defaultValue">Default value to ignore.</param>
+        /// <returns>
+        /// Wrapped value.
+        /// </returns>
+        public T OrDefault(T defaultValue)
+        {
+            return Value;
+        }
+
+        /// <summary>
+        /// Compare this instance with another option value
+        /// </summary>
+        /// <param name="other">Other value to compare with</param>
+        /// <returns>True if the two values are equal (value semantics).</returns>
+        public bool Equals(IOption<T> other)
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            var value = Value;
+            return other.Match(() => false, v => Equals(v, value));
+        }
+    }
+
+    /// <summary>
+    /// Represents the abscence of an optional value
+    /// </summary>
+    /// <typeparam name="T">Type of value we don't have.</typeparam>
+    public sealed class None<T> : IOption<T>
+    {
+        /// <summary>
+        /// Compare this instance with another option value
+        /// </summary>
+        /// <param name="other">Other value to compare with</param>
+        /// <returns>True if the two values are equal (value semantics).</returns>
+        public bool Equals(IOption<T> other)
+        {
+            return other?.GetType() == GetType();
```

Comment:
perhaps simpler as `other is None<T>`?

Reply:
note that this notion of equality is a bit different to the one on `Some<T>`? like I can write my own `GeorgeNone` and this won't say that it's equal, but the `Some<T>` one I could write a `GeorgeSome` and have it work...

Reply:
I've sealed the classes, so no you can't. ;-)